iceoryx_posh/source/runtime/service_discovery.cpp

```diff
@@ -21,97 +21,53 @@ namespace iox
-    bool isInstanceWildcard = !instance;
-    bool isEventWildcard = !event;
+    // Copy the new service registry if has changed
+    m_serviceRegistrySubscriber.take().and_then([&](popo::Sample<const roudi::ServiceRegistry>& serviceRegistrySample) {
+        m_serviceRegistry = *serviceRegistrySample;
+    });
 
-    if (!isServiceWildcard)
-    {
-        serviceString = service.value();
-    }
-    if (!isInstanceWildcard)
-    {
-        instanceString = instance.value();
-    }
-    if (!isEventWildcard)
-    {
-        eventString = event.value();
-    }
-
-    IpcMessage sendBuffer;
-    sendBuffer << IpcMessageTypeToString(IpcMessageType::FIND_SERVICE) << PoshRuntime::getInstance().getInstanceName()
-               << cxx::convert::toString(isServiceWildcard) << serviceString
-               << cxx::convert::toString(isInstanceWildcard) << instanceString
-               << cxx::convert::toString(isEventWildcard) << eventString;
-
-    IpcMessage requestResponse;
+    roudi::ServiceRegistry::ServiceDescriptionVector_t tempSearchResult;
+    m_serviceRegistry.find(tempSearchResult, service, instance, event);
 
-    if (!PoshRuntime::getInstance().sendRequestToRouDi(sendBuffer, requestResponse))
+    ServiceContainer searchResult;
+    for (auto& service : tempSearchResult)
     {
-        LogError() << "Could not send FIND_SERVICE request to RouDi\n";
-        errorHandler(Error::kIPC_INTERFACE__REG_UNABLE_TO_WRITE_TO_ROUDI_CHANNEL, nullptr, ErrorLevel::MODERATE);
-        return cxx::error<FindServiceError>(FindServiceError::UNABLE_TO_WRITE_TO_ROUDI_CHANNEL);
+        searchResult.push_back(service.serviceDescription);
     }
 
-    ServiceContainer serviceContainer;
-    uint32_t numberOfElements = requestResponse.getNumberOfElements();
-    uint32_t capacity = static_cast<uint32_t>(serviceContainer.capacity());
+    return searchResult;
+}
 
-    // Limit the services (max value is the capacity of serviceContainer)
-    uint32_t numberOfServices = algorithm::min(capacity, numberOfElements);
-    for (uint32_t i = 0U; i < numberOfServices; ++i)
+void ServiceDiscovery::enableEvent(popo::TriggerHandle&& triggerHandle, const ServiceDiscoveryEvent event) noexcept
+{
+    if (event == ServiceDiscoveryEvent::SERVICE_REGISTRY_HAS_CHANGED)
     {
-        auto deserializationResult =
-            capro::ServiceDescription::deserialize(cxx::Serialization(requestResponse.getElementAtIndex(i)));
-        if (deserializationResult.has_error())
-        {
-            return cxx::error<FindServiceError>(FindServiceError::DESERIALIZATION_FAILED);
-        }
-        serviceContainer.push_back(deserializationResult.value());
+        m_serviceRegistrySubscriber.enableEvent(std::move(triggerHandle), popo::SubscriberEvent::DATA_RECEIVED);
     }
+}
 
-    if (numberOfElements > capacity)
+void ServiceDiscovery::disableEvent(const ServiceDiscoveryEvent event) noexcept
+{
+    if (event == ServiceDiscoveryEvent::SERVICE_REGISTRY_HAS_CHANGED)
```

Comment:
Same here, please add switch case where the default case calls the error handler.